sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/pointer/CommonPointerLibraries.java

```diff
@@ -119,6 +146,92 @@ static boolean isMemberModifiable(LLVMPointerImpl receiver, String ident,
+            return clazz.findMethod(ident) != null;
+        }
+        return false;
+    }
+
+    @ExportMessage
+    static class InvokeMember {
+        /**
+         * @param member
+         * @param context
+         * @param clazz
+         * @param method
+         * @param argCount
+         * @param methodName
+         * @param llvmFunction
+         * @see InteropLibrary#invokeMember(Object, String, Object[])
+         */
+        @Specialization(guards = {"asClazz(receiver)==clazz", "member.equals(methodName)", "argCount==arguments.length"})
+        static Object doCached(LLVMPointerImpl receiver, String member, Object[] arguments,
+                        @CachedContext(LLVMLanguage.class) LLVMContext context, @CachedLibrary(limit = "5") InteropLibrary interop,
+                        @Cached(value = "asClazz(receiver)") LLVMInteropType.Clazz clazz,
+                        @Cached(value = "clazz.findMethodByArguments(receiver, member, arguments)") Method method,
+                        @Cached(value = "arguments.length") int argCount,
+                        @Cached(value = "method.getName()") String methodName,
+                        @Cached(value = "getLLVMFunction(context, method, clazz, member)") LLVMFunction llvmFunction,
+                        @Cached(value = "create(llvmFunction)") LLVMAccessSymbolNode accessSymbolNode)
+                        throws UnsupportedMessageException, ArityException, UnsupportedTypeException {
+            Object[] newArguments = addSelfObject(receiver, arguments);
+            return interop.execute(accessSymbolNode.execute(), newArguments);
+        }
+
+        @Specialization(replaces = "doCached")
+        static Object doResolve(LLVMPointerImpl receiver, String member, Object[] arguments,
+                        @CachedContext(LLVMLanguage.class) LLVMContext context, @CachedLibrary(limit = "5") InteropLibrary interop,
+                        @Cached LLVMDynAccessSymbolNode dynAccessSymbolNode)
+                        throws UnsupportedMessageException, ArityException, UnsupportedTypeException, UnknownIdentifierException {
+            Object[] newArguments = addSelfObject(receiver, arguments);
+            LLVMInteropType.Clazz newClazz = asClazz(receiver);
+            Method newMethod = newClazz.findMethodByArguments(receiver, member, arguments);
+            LLVMFunction newLLVMFunction = getLLVMFunction(context, newMethod, newClazz, member);
+            Object newReceiver = dynAccessSymbolNode.execute(newLLVMFunction);
+            return interop.execute(newReceiver, newArguments);
+        }
+    }
+
+    static LLVMInteropType.Clazz asClazz(LLVMPointerImpl receiver) throws UnsupportedTypeException {
+        LLVMInteropType type = receiver.getExportType();
+        if (!(type instanceof LLVMInteropType.Clazz)) {
+            throw UnsupportedTypeException.create(new Object[]{receiver}, receiver + " cannot be casted to LLVMInteropType.Clazz");
```

Comment:
String concatenation is not safe on the fast path (internally this creates quite a lot of code). In this particular case, I'd just use a simple string (i.e. literally `"Receiver cannot be cast to LLVMInteropType.Clazz"`, or maybe better `"Receiver type is not a class."`).

Reply:
(found by our internal gate)